addons/account_tax_python/models/account_tax.py

```diff
@@ -1,94 +1,197 @@
         # EXTENDS 'account'
         field_names = super()._eval_taxes_computation_prepare_product_fields(taxes_data)
-        Product = self.env['product.product']
         for tax_data in taxes_data:
             if tax_data['amount_type'] == 'code':
-                tax = self.browse(tax_data['id'])
-                for formula in ((tax.python_applicable or '').strip(), (tax.python_compute or '').strip()):
-                    groups = REGEX_FORMULA_OBJECT.findall(formula)
-                    if groups:
-                        for group in groups:
-                            field_name = group[1]
-                            if field_name in Product:
-                                field_names.add(field_name)
-
+                field_names.update(tax_data['_product_fields'])
         return field_names
 
+    def _decode_formula(self, formula):
+        """ Decode the formula and extract relevant values from it.
+
+        :param formula: The value of the 'formula' field.
+        """
+        self.ensure_one()
+
+        if self.amount_type != 'code':
+            return {}
+
+        formula = (formula or '0.0').strip()
+        results = {
+            '_js_formula': formula
+                .replace('and', '&&')
+                .replace('or', '||')
+                .replace('None', 'undefined'),
+            '_py_formula': formula,
+        }
+        product_fields = set()
+
+        groups = re.findall(r'((?:product\.)(?P<field>\w+))+', formula) or []
+        Product = self.env['product.product']
+        for group in groups:
+            field_name = group[1]
+            if field_name in Product:
+                product_fields.add(field_name)
+                results['_py_formula'] = results['_py_formula'].replace(f"product.{field_name}", f"product['{field_name}']")
+
+        results['_product_fields'] = list(product_fields)
+        return results
+
+    @api.model
+    def _check_formula(self, tax_data):
+        """ Check the formula is passing the minimum check to ensure the compatibility between both evaluation
+        in python & javascript.
+
+        :param tax_data: The values returned by '_prepare_dict_for_taxes_computation'.
+        """
+        def get_number_size(formula, i):
+            starting_i = i
+            seen_separator = False
+            while i < len(formula):
+                if formula[i].isnumeric():
+                    i += 1
+                elif formula[i] == '.' and (i - starting_i) > 0 and not seen_separator:
+                    i += 1
+                    seen_separator = True
+                else:
+                    break
+            return i - starting_i
+
+        allowed_tokens = FORMULA_ALLOWED_TOKENS.union(f"product['{field_name}']" for field_name in tax_data['_product_fields'])
+        formula = tax_data['_py_formula']
+
+        def raise_malformed():
+            raise ValidationError(_("Malformed formula '%s' at %s", formula, i))
```

Comment:
please use placeholders when there's more than 1 variable, as in some languages the order of words might switch